crates/ra_hir_def/src/nameres/collector.rs

```diff
@@ -407,7 +414,18 @@ where
         import: Option<LocalImportId>,
         resolutions: &[(Name, Resolution)],
     ) {
-        self.update_recursive(module_id, import, resolutions, 0)
+        self.update_recursive(module_id, import, resolutions, 0);
+
+        // update alias
+        let alias_resolutions: Vec<_> = resolutions
+            .iter()
+            .filter_map(|(name, res)| {
+                let alias = self.def_map.modules[module_id].aliases.get(name)?;
+                Some((alias.clone(), res.clone()))
+            })
+            .collect();
+
```

Comment:
Hm, I am feeling uneasy about this approach. Seems like `Name -> Name` mapping errases essential information (like from what import the name initially came). I imagine `use foo::a as b; use bar::a as b` might actually be legal if `a`'s live in different namespaces? 

And I think we actually have all the required infrastructure in place to fix this properly? Basically, an `use foo as bar` import should be marked unresolved, and, when we add `foo` to the module namespace, on the next iteration of the `loop { }` we should also add `bar`. It seems like we need to fix some bug in existing infra, rather than add something specific for aliases tracking.

Reply:
> And I think we actually have all the required infrastructure in place to fix this properly? Basically, an use foo as bar import should be marked unresolved.

Um.. imagine this example: 
```rust
mod m {
    macro_rules! _foo {
        ($x:ident) => { type $x = u64; }
    }
    pub(crate) use _foo as foo;
}

macro_rules! foo {
    () => {}
}

m::foo!(foo);
use foo as bar;   // <-----
```
The order of resolving will be: 
`mod m`
`macro_rules _foo 
`use _foo as foo`
`macro_rules foo`
`use foo as bar  // foo is macro_rules! foo`
`m:foo!(foo)`
`type foo = u64`

Reply:
> Hm, I am feeling uneasy about this approach. Seems like Name -> Name mapping errases essential information (like from what import the name initially came). I imagine use foo::a as b; use bar::a as b might actually be legal if a's live in different namespaces?

Yes, I agree, I didn't think of this problem.  It is possible to store more information in the alias ?

Reply:
Ah, so the problem is that we end up with `bar` only as `macro`, and miss `bar` as a `type`, right? 

So looks like the problem is that we remove `use foo as bar` from the set of unresolved imports when it resolves in *any* namespace, while we should only do this if it resolves in *all* namespaces. I wonder if we construct a test case here which doens't involve any macros? something like this

```rust
type T = ();

use self::T as  Y;

use m::*;

mod m {
  pub const T: () = ();
}

fn main() { let _: Y = Y; }
```

Looks like rust-analyzer indeed fails to see that `Y` is a value here:


![image](https://user-images.githubusercontent.com/1711539/69974210-1516a400-1525-11ea-8721-4cdac6ea7366.png)

Reply:
yeah, one doesn't need aliases to construct a similar problem:

![image](https://user-images.githubusercontent.com/1711539/69974428-89514780-1525-11ea-930e-73f26d453f05.png)

![image](https://user-images.githubusercontent.com/1711539/69974446-91a98280-1525-11ea-843b-16f46c4ffaea.png)

Reply:
Yes indeed. The problem here is the `use alias`. My approach is to fix it is mark the alias and rewrite to it.  OTOH, we can mark it to unresolved, but the problem is, how can we know the resolve `loop` should be ended ?

Reply:
> yeah, one doesn't need aliases to construct a similar problem:

Okay, my alias fix won't work in your case. :)

Reply:
Just checked that IntelliJ also doesn't handle the above example :weary: